src/v/pandaproxy/schema_registry/avro.cc

```diff
@@ -530,66 +530,79 @@ ss::sstring avro_schema_definition::name() const {
 
 class collected_schema {
 public:
-    bool contains(const ss::sstring& name) const {
-        return _names.contains(name);
+    bool contains(const avro::Name& name) const {
+        return _schemas.contains(name);
     }
-    bool insert(ss::sstring name, schema_definition def) {
-        bool inserted = _names.insert(std::move(name)).second;
-        if (inserted) {
-            _schemas.push_back(std::move(def).raw());
-        }
+
+    bool insert(avro::Name name, avro::ValidSchema schema) {
+        auto [it, inserted] = _schemas.try_emplace(
+          std::move(name), std::move(schema));
         return inserted;
     }
-    schema_definition::raw_string flatten() && {
-        iobuf out;
-        for (auto& s : _schemas) {
-            out.append(std::move(s));
-            out.append("\n", 1);
-        }
-        return schema_definition::raw_string{std::move(out)};
+
+    const std::map<avro::Name, avro::ValidSchema>& as_named_references() const {
+        return _schemas;
+    }
+
+    void merge(collected_schema other) {
+        _schemas.merge(std::move(other._schemas));
     }
 
 private:
-    absl::flat_hash_set<ss::sstring> _names;
-    std::vector<schema_definition::raw_string> _schemas;
+    std::map<avro::Name, avro::ValidSchema> _schemas;
```

Comment:
I'd be tempted to change `collected_schema` to a using declaration now that it's just a thin wrapper over `std::map`.